app/models/zone.rb

```diff
@@ -5,22 +5,31 @@ class Zone < ApplicationRecord
   has_many :labels, -> { order(designated_on: :desc) }, class_name: 'ZoneLabel', inverse_of: :zone
   has_and_belongs_to_many :procedures, -> { order(published_at: :desc) }, inverse_of: :zone
 
+  LAST_ZONE = 'Autre'.freeze
+
   def current_label
     labels.where.not(name: 'Non attribué').first.name
   end
 
   def label_at(date)
-    label = labels.where('designated_on < ?', date)&.first || labels.last
-    label.name
+    labels.where('designated_on < ?', date).first&.name || labels.last&.name
   end
 
   def available_at?(date)
     label_at(date) != 'Non attribué'
   end
 
   def self.available_at(date, without_zones = [])
-    (Zone.all - without_zones).filter { |zone| zone.available_at?(date) }.sort_by { |zone| zone.label_at(date) }
-      .map do |zone|
+    zones = Zone.includes(:labels).where.not(id: without_zones).to_a
+
+    zone_labels = zones.each_with_object({}) do |zone, labels|
+      label = zone.labels.filter { |l| l.designated_on < date }.max_by(&:designated_on)
+      labels[zone.id] = label&.name || ''
+    end
+
+    available_zones = zones.filter { |zone| zone_labels[zone.id] != 'Non attribué' }
+
+    available_zones.map do |zone|
```

Comment:
Alors, je l'ai pas fait tournée mais voici ma proposition :

d'abords on peut passer `label_at(date)` en privée.
ensuite, on peut la redéfinir pour qu'elle ne fasse plus d'AR

```
def label_at(date)
  labels
    .sort { _1.designated_on }.reverse
    .filter { _1.designated_on < date}.first&.name || labels.last&.name
end
```

puis simplifier ce code

```suggestion
    Zone.includes(:labels).where.not(id: without_zones)
      .filter { _1available_at?(date) }
      .map do |zone|
```

WDYT ?

Reply:
Bien vu pour éviter les allers retours ! Du coup j'ai regardé plus attentivement avec rack profiler ! 

Pour éviter les aller retours j'ai trouvé ceci, t'en pense quoi ? On se passe du label_at totalement du coup.

On passe de 58 requetes sql à 6 !

```ruby
  def self.available_at(date, without_zones = [])
    # Charger toutes les zones qui ne sont pas dans without_zones
    zones = Zone.where.not(id: without_zones).order(:acronym).to_a

    # Charger tous les labels pertinents (ceux désignés avant la date donnée) en une seule requête
    labels_by_zone = ZoneLabel
      .where('designated_on <= ?', date)
      .where(zone_id: zones.map(&:id))
      .order(designated_on: :desc)
      .group_by(&:zone_id)

    # Préparer la liste des zones avec leurs labels
    zones.map do |zone|
      label = labels_by_zone[zone.id]&.first&.name
      next if label.nil? || label == 'Non attribué' # Exclure les zones sans label ou avec "Non attribué"
      LabelModel.new(id: zone.id, label: label)
    end.compact
  end
```

Ce qui nous permet de ne pas toucher à la méthode "label_at". Car là en l'état c'est une requete sql donc c'est pas mal.

enchainer les manip ça me parait un peu complexe non ?
faire un .sort -> O(n log n)
puis un .reverse -> O(n)
puis un .filter -> O(n)
ça nous fait enchainer du lourd niveau ressource je pense.

Une petite requête SQL avec le where et hop ! Qu'en penses tu ?